botocore/utils.py

```diff
@@ -794,5 +800,117 @@ def _switch_hosts(request, new_endpoint, use_new_scheme=True):
+    logger.debug('Updating URI from %s to %s' % (
+        original_endpoint, final_endpoint))
+    return final_endpoint
+
+
+class S3RegionRedirector(object):
+    def __init__(self, endpoint_bridge, cache=None):
+        self._endpoint_resolver = endpoint_bridge
+        self._cache = cache or {}
+        self._cache_lock = threading.Lock()
+
+    def _add_bucket_to_cache(self, bucket, signing_context):
+        with self._cache_lock:
+            self._cache[bucket] = signing_context
+
+    def _get_bucket_from_cache(self, bucket):
+        with self._cache_lock:
+            signing_context = self._cache.get(bucket, None)
+        return signing_context
+
+    def register(self, emitter):
+        emitter.register('needs-retry.s3', self.redirect_from_error)
+        emitter.register('before-call.s3', self.set_request_url)
+        emitter.register('before-parameter-build.s3', self.redirect_from_cache)
+
+    def redirect_from_error(self, request_dict, response, **kwargs):
+        """
+        An S3 request sent to the wrong region will return an error that
+        contains the endpoint the request should be sent to. This handler
+        will add the redirect information to the signing context and then
+        redirect the request.
+        """
+        if response is None:
+            # This could be none if there was a ConnectionError or other
+            # transport error.
+            return
+
+        error = response[1].get('Error', {})
+        if error.get('Code') != 'PermanentRedirect':
+            return
+
+        if 'context' not in request_dict:
+            request_dict['context'] = {}
+
+        bucket = error['Bucket']
+        client_region = request_dict['context'].get('client_region')
+        new_region = self.get_region_from_response(response)
+
+        if new_region is None:
+            logger.debug(
+                "S3 client configured for region %s but the bucket %s is not "
+                "in that region and the proper region could not be "
+                "automatically determined." % (client_region, bucket))
+            return
+
+        logger.debug(
+            "S3 client configured for region %s but the bucket %s is in region"
+            " %s; Please configure the proper region to avoid multiple "
+            "unnecessary redirects and signing attempts." % (
+                client_region, bucket, new_region))
+
+        endpoint = self._endpoint_resolver.resolve('s3', new_region)
+        endpoint = endpoint['endpoint_url']
+
+        signing_context = {
+            'region': new_region,
+            'bucket': bucket,
+            'endpoint': endpoint
+        }
+        request_dict['context']['signing'] = signing_context
+
+        self._add_bucket_to_cache(bucket, signing_context)
+        self.set_request_url(request_dict, request_dict['context'])
+
+        # Return 0 so it doesn't wait to retry
+        return 0
+
+    def get_region_from_response(self, response):
```

Comment:
This name doesn't match with the description.  This isn't just getting the region from the response.  It will potentially make network calls if necessary.